rtl8752h porting

zophai-liu opened the pull request:
- Add necessary files for rtl8752h. The hello world sample can be built.
- cherry-pick drivers from [pull/70](https://github.com/rtkconnectivity/zephyr/pull/70/commits)
- Add rtk init code involved vectors and osif

YuzhuoLiuRTK committed:
drivers: add clock_control driver for rtl8752h

support clock control for rtl8752h

Signed-off-by: Yuzhuo Liu <yuzhuo_liu@realsil.com.cn>

YuzhuoLiuRTK committed:
drivers: add pinctrl driver for rtl8752h

support pinctrl for rtl8752h

Signed-off-by: Yuzhuo Liu <yuzhuo_liu@realsil.com.cn>

YuzhuoLiuRTK committed:
drivers: add serial driver for rtl8752h

support serial for rtl8752h

Signed-off-by: Yuzhuo Liu <yuzhuo_liu@realsil.com.cn>

zophai-liu committed:
soc: arm: realtek_bee: add rtl8752h

add basic files for rtl8752h

Signed-off-by: Zophai Liu <zophai_liu@realsil.com.cn>

zophai-liu committed:
dts: arm: realtek: add rtl8752h

add necessary node for rtl8752h

Signed-off-by: Zophai Liu <zophai_liu@realsil.com.cn>

zophai-liu committed:
boards: arm: add rtl8752htv_evb

add rtl8752htv_evb support

Signed-off-by: Zophai Liu <zophai_liu@realsil.com.cn>

YuzhuoLiuRTK committed:
boards: add console uart for rtl8752htv_evb

1.add clock control, pinctrl, uart, dma, gpio in rtl8752h.dtsi
2.support console for rtl8752htv_evb

Signed-off-by: Yuzhuo Liu <yuzhuo_liu@realsil.com.cn>

YuzhuoLiuRTK committed:
drivers: add dma driver for rtl8752h

support dma for rtl8752h

Signed-off-by: Yuzhuo Liu <yuzhuo_liu@realsil.com.cn>

YuzhuoLiuRTK committed:
tests: add uart test for rtl8752htv_evb

support uart_basic_api test and uart_async_api test for rtl8752htv_evb

Signed-off-by: Yuzhuo Liu <yuzhuo_liu@realsil.com.cn>

YuzhuoLiuRTK committed:
tests: add dma loop test for rtl8752htv_evb

add rtl8752htv_evb.config and rtl8752htv_evb.overlay files for dma
loop_transfer test

Signed-off-by: Yuzhuo Liu <yuzhuo_liu@realsil.com.cn>

YuzhuoLiuRTK committed:
drivers: add gpio driver for rtl8752h

support gpio for rtl8752h

Signed-off-by: Yuzhuo Liu <yuzhuo_liu@realsil.com.cn>

YuzhuoLiuRTK committed:
tests: add gpio basic_api test for rtl8752htv_evb

add rtl8752htv_evb.overlay files for gpio basic_api test

Signed-off-by: Yuzhuo Liu <yuzhuo_liu@realsil.com.cn>

zophai-liu committed:
soc: add rom init for rtl8752h

add rtk init, such as osif, vectors recovery

Signed-off-by: Zophai Liu <zophai_liu@realsil.com.cn>

zophai-liu committed:
driver: flash: add rtl8752h

add rtl875h flash rw driver,
not include flash init

Signed-off-by: Zophai Liu <zophai_liu@realsil.com.cn>

zophai-liu committed:
dts: arm: realtek: update rtl8752h ram region

Expand zephyr sram with shrinking heap to meet
BLE project requirements

Signed-off-by: Zophai Liu <zophai_liu@realsil.com.cn>

zophai-liu committed:
drivers: bluetooth: hci: add rtl8752h

rtl8752h use hci_rtl87x2g.c temporarily

Signed-off-by: Zophai Liu <zophai_liu@realsil.com.cn>

zophai-liu committed:
boards: arm: rtl8752htv_evb: add BT defconfig

This commit adds BT default config for rtl8752htv_evb

Signed-off-by: Zophai Liu <zophai_liu@realsil.com.cn>

zophai-liu committed:
drivers: entropy: add rtl8752h

This commit adds entropy driver for rtl8752h.

Signed-off-by: Zophai Liu <zophai_liu@realsil.com.cn>

zophai-liu committed:
dts: arm: realtek: update rtl8752h flash allocation

This commit updates rtl8752h flash allocation for
app_partition and storage_partition.

Signed-off-by: Zophai Liu <zophai_liu@realsil.com.cn>

ZhiyuanTang17 commented on boards/arm/rtl8752htv_evb/rtl8752htv_evb.yaml:
```diff
@@ -0,0 +1,31 @@
+  - zephyr
+  - gnuarmemb
+  - xtools
+supported:
```
这边support的外设再review下，有些bee3plus上应该没有。

zophai-liu replied:
ok

ZhiyuanTang17 commented on boards/arm/rtl8752htv_evb/rtl8752htv_evb_defconfig:
```diff
@@ -0,0 +1,24 @@
+CONFIG_USE_DT_CODE_PARTITION=y
+
+# For Debug
+CONFIG_DEBUG_OPTIMIZATIONS=y
```
这个默认要开吗？

zophai-liu replied:
后面再关

ZhiyuanTang17 commented on drivers/clock_control/clock_control_rtl8752h.c:
```diff
@@ -0,0 +1,97 @@
+/*
+ * Copyright(c) 2020, Realtek Semiconductor Corporation.
```
license注释里的时间统一改成2024

zophai-liu replied:
ok

ZhiyuanTang17 commented on soc/arm/realtek_bee/rtl8752h/soc.c:
```diff
@@ -0,0 +1,348 @@
+	if (flash_sem == NULL) {
+		os_sem_create(&flash_sem, "ftl_sem", 1, 1);
+	}
+}
```
这个sem应该不需要了？因为zephyr不用ftl了。

ZhiyuanTang17 commented on soc/arm/realtek_bee/rtl8752h/soc.c:
```diff
@@ -0,0 +1,348 @@
+
+	rtk_task_init();
+
+	if (flash_nor_get_exist(FLASH_NOR_IDX_SPIC0) != FLASH_NOR_EXIST_NONE) {
+		if (flash_nor_load_query_info(FLASH_NOR_IDX_SPIC0) == FLASH_NOR_RET_SUCCESS) {
+			/* apply SW Block Protect */
+			if (boot_cfg.flash_setting.bp_enable) {
+				/**
+				 * set flash default block protect level depend on
+				 * different flash id and different flash layout
+				 */
+				boot_cfg.flash_setting.bp_lv = flash_nor_get_default_bp_lv();
+			} else {
+				boot_cfg.flash_setting.bp_lv = 0;
+			}
+
+			if (flash_nor_set_tb_bit(FLASH_NOR_IDX_SPIC0, 1) == FLASH_NOR_RET_SUCCESS &&
+			flash_nor_set_bp_lv(FLASH_NOR_IDX_SPIC0,
+			boot_cfg.flash_setting.bp_lv) == FLASH_NOR_RET_SUCCESS){
+				FLASH_PRINT_INFO1("Flash BP Lv = %d", boot_cfg.flash_setting.bp_lv);
+			} else {
+				FLASH_PRINT_INFO0("Flash BP fail!");
+			}
+		}
+		flash_nor_dump_flash_info();
+	}
```
这部分看看之后能不能放到flash driver里的init里去。

zophai-liu committed:
soc: arm: realtek_bee: change IRQ restore method

Fix bug: z_arm_exc_exit not executed after irq recovery.
Use irq_connect_dynamic to restore rom project irqs.

Signed-off-by: Zophai Liu <zophai_liu@realsil.com.cn>

zophai-liu committed:
boards: arm: rtl8752htv_evb: update app_partition

Adjust app_partition due to increased stack patch space.

Signed-off-by: Zophai Liu <zophai_liu@realsil.com.cn>

zophai-liu committed:
soc: arm: realtek_bee: set lowerstack entry at early level

Change the entry of lowerstack from `POST_KERNEL` to `EARLY`

Signed-off-by: Zophai Liu <zophai_liu@realsil.com.cn>

zophai-liu committed:
soc: rtl8752h: fix orphan trace section

Add trace to ld scripts

Signed-off-by: Zophai Liu <zophai_liu@realsil.com.cn>

zophai-liu committed:
driver: flash: rtl8752h 4bit mode

This commit supports changing flash bit mode to speed up.

Signed-off-by: Zophai Liu <zophai_liu@realsil.com.cn>

zophai-liu committed:
drivers: bluetooth: add BT config for realtek hci

This commit adds CONFIG_BT_RTK_HCI_H2C_POOL_SIZE for realtek hci.

Signed-off-by: Zophai Liu <zophai_liu@realsil.com.cn>

zophai-liu committed:
boards: rtl8752htv_evb: add controller stack size defconfig

This commit adds default value for
CONFIG_REALTEK_BT_CONTROLLER_STACK_SIZE.

Signed-off-by: Zophai Liu <zophai_liu@realsil.com.cn>

zophai-liu committed:
soc: rtl8752h: fix irq restore bug

This commits fixs irq restore bug. When an IRQ has
been enabled, configuring it with irq_connect_dynamic
would trigger an assert.

Signed-off-by: Zophai Liu <zophai_liu@realsil.com.cn>

zophai-liu committed:
soc: rtl8752h: fix build warnings

This commit fixs build warnings.

Signed-off-by: Zophai Liu <zophai_liu@realsil.com.cn>

zophai-liu committed:
board: arm: add rtl8752h_dongle

This commits add rtl8752h_dongle board and RTL8752HKF

Signed-off-by: Zophai Liu <zophai_liu@realsil.com.cn>

zophai-liu committed:
driver: flash: reorder "define" and "include"

This commit reorders "define" and "include" to fix
build error in handling devicetree_generated.h

Signed-off-by: Zophai Liu <zophai_liu@realsil.com.cn>
